src/jsonrpc-virtual-machine.cpp

```diff
@@ -48,90 +57,153 @@ std::string jsonrpc_post_data(const std::string &method, const std::tuple<Ts...>
+        const boost::asio::ip::tcp::no_delay no_delay_option(true);
+        stream.socket().set_option(no_delay_option);
+
+        // Minimize socket close time by setting the linger time to 0.
+        // On MacOS, it avoids accumulating socket in TIME_WAIT state, after rapid successive requests,
+        // which can consume all available ports.
+        // It is safe to do this because it is the client who decides to close the connection,
+        // after all data is received.
+        const boost::asio::socket_base::linger linger_option(true, 0);
+        stream.socket().set_option(linger_option);
+
+        // Enable keep alive TCP option for keep alive HTTP connection
+        if (keep_alive) {
+            const boost::asio::socket_base::keep_alive keep_alive_option(true);
+            stream.socket().set_option(keep_alive_option);
+        }
     }
-    if (data.status_code != "200") {
-        throw std::runtime_error("http error: "s + data.reason_phrase + " (code "s + data.status_code + ")"s);
+
+    try {
+        // Set up a HTTP request message
+        http::request<http::string_body> req;
+        req.method(http::verb::post); // POST
+        req.version(11);              // Only HTTP 1.1 support keep alive connections
+        req.target(uri);
+        req.keep_alive(keep_alive);
+        req.set(http::field::host, host);
+        req.set(http::field::content_type, "application/json");
+        req.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+        req.body() = post_data;
+        req.prepare_payload();
+
+        // Send the HTTP request
+        while (true) {
+            beast::error_code ec;
+            http::write(stream, req, ec);
+            if (!ec) { // Success
+                break;
+            } else if (ec == asio::error::interrupted) {
+                // Retry the operation during interrupts (SIGINT/SIGTERM),
+                // otherwise we may leave dead zombies processes during fork requests.
+            } else { // Unexpected error
+                throw beast::system_error(ec);
+            }
+        }
+
+        // Set up HTTP response parser
+        beast::flat_buffer buffer;
+        http::response_parser<http::string_body> res_parser;
+        res_parser.eager(true);
+        res_parser.body_limit(16777216U); // can receive up to 16MB
+
+        // Receive the HTTP response
+        while (true) {
+            beast::error_code ec;
+            http::read(stream, buffer, res_parser, ec);
+            if (!ec) { // Success
+                break;
+            } else if (ec == asio::error::interrupted) {
+                // Retry the operation during interrupts (SIGINT/SIGTERM),
+                // otherwise we may leave dead zombies processes during fork requests.
+            } else { // Unexpected error
+                throw beast::system_error(ec);
+            }
+        }
+
+        http::response<http::string_body> res = res_parser.get();
```

Comment:
Is this a copy?

Reply:
A shallow copy (the message buffer is not copied), but I will use `const&` .

Reply:
Done.